lnc/lnc.go

```diff
@@ -0,0 +1,331 @@
+		localStatic, err := btcec.NewPrivateKey()
+		if err != nil {
+			return fmt.Errorf("unable to generate local static "+
+				"key: %w", err)
+		}
+
+		session.LocalStaticPrivKey = localStatic
+
+		err = l.store.AddSession(ctxb, session)
+		if err != nil {
+			return fmt.Errorf("unable to add new session: %w", err)
+		}
+
+	case err != nil:
+		return fmt.Errorf("unable to get session(%v): %w",
+			session.PassphraseEntropy, err)
+
+	default:
+		if session.Label != dbSession.Label {
+			return fmt.Errorf("the lnc session already exists "+
+				"with a different label: %v", session.Label)
+		}
+
+		session = dbSession
+	}
+
+	l.session = session
+
+	conn, err := l.newConn(session)
+	if err != nil {
+		return err
+	}
+
+	l.conn = conn
+
+	return nil
+}
+
+// CloseConn closes the connection with the remote node.
+func (l *NodeConn) CloseConn() error {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	if l.conn == nil {
+		return fmt.Errorf("connection not open")
+	}
+
+	err := l.conn.Close()
+	if err != nil {
+		return fmt.Errorf("unable to close connection: %w", err)
+	}
+
+	l.conn = nil
+	l.session = nil
+	l.macStr = ""
+
+	return nil
+}
+
+// Stop closes the connection with the remote node if it is open.
+func (l *NodeConn) Stop() error {
+	if l.conn != nil {
+		return l.CloseConn()
+	}
+
+	return nil
+}
+
+// Client returns the gRPC client to the remote node.
+func (l *NodeConn) Client() (lnrpc.LightningClient, error) {
+	if l.conn == nil {
+		return nil, fmt.Errorf("connection not open")
+	}
+
+	return l.conn.client, nil
+}
+
+// Conn returns the context that needs to be used whenever the internal Client
+// is used.
+func (l *NodeConn) CtxFunc() context.Context {
+	ctx := context.Background()
+	return metadata.AppendToOutgoingContext(ctx, HeaderMacaroon, l.macStr)
+}
+
+// onRemoteStatic is called when the remote static key is received.
+//
+// NOTE: this function is a callback to be used by the mailbox package during
+// the mailbox.NewConnData call.
+func (l *NodeConn) onRemoteStatic(key *btcec.PublicKey) error {
+	ctxt, cancel := context.WithTimeout(
+		context.Background(), time.Second*10,
+	)
+	defer cancel()
+
+	remoteKey := key.SerializeCompressed()
+
+	err := l.store.SetRemotePubKey(ctxt, l.session.Label, remoteKey)
+	if err != nil {
+		log.Errorf("unable to set remote pub key for session(%v): %w",
+			l.session.Label, err)
+	}
+
+	return err
+}
+
+// onAuthData is called when the auth data is received.
+//
+// NOTE: this function is a callback to be used by the mailbox package during
+// the mailbox.NewConnData call.
+func (l *NodeConn) onAuthData(data []byte) error {
+	mac, err := extractMacaroon(data)
+	if err != nil {
+		log.Errorf("unable to extract macaroon for session(%v): %w",
+			l.session.Label, err)
+
+		return err
+	}
+
+	macBytes, err := mac.MarshalBinary()
+	if err != nil {
+		log.Errorf("unable to marshal macaroon for session(%v): %w",
+			l.session.Label, err)
+
+		return err
+	}
+
+	l.macStr = hex.EncodeToString(macBytes)
+
```

Comment:
should we verify that the macaroon we get actually does have the permissions that aperture will need to operate?

Reply:
Do we have methods exported from the `lnd` macaroons package or aperture to easily sanity check this?

Reply:
TODO added :+1: